src/include/codegen/util/hash_table.h

```diff
@@ -0,0 +1,327 @@
+   * @param key The key to store in the table
+   * @param value The value to store in the value
+   */
+  template <typename Key, typename Value>
+  void TypedInsertLazy(uint64_t hash, const Key &key, const Value &value);
+
+  /**
+   * Insert a key-value pair into the hash table. This function is used mostly
+   * for testing since we specialize insertions in generated code.
+   *
+   * @param hash The hash value of the key
+   * @param key The key to store in the table
+   * @param value The value to store in the value
+   */
+  template <typename Key, typename Value>
+  void TypedInsert(uint64_t hash, const Key &key, const Value &value);
+
+  /**
+   * Probe a key in the hash table. This function is used mostly for testing.
+   * Probing is completely specialized in generated code.
+   *
+   * @param hash The hash value of the key
+   * @param key The key to probe in the table
+   * @param[out] value The value associated with the key in the table
+   * @return True if a value was found. False otherwise.
+   */
+  template <typename Key, typename Value>
+  bool TypedProbe(uint64_t hash, const Key &key,
+                  std::function<void(const Value &)> &consumer);
+
+  //////////////////////////////////////////////////////////////////////////////
+  ///
+  /// Helper Classes
+  ///
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * An entry in the hash table that stores a key and value. Entries are
+   * variable-sized structs. In memory, the key and value bytes are stored
+   * contiguously after the next pointer. The layout is thus:
+   *
+   * +--------------+---------------+-------------------+--------------------+
+   * |  hash value  | next pointer  | sizeof(Key) bytes | sizeof(Val) bytes) |
+   * +--------------+---------------+-------------------+--------------------+
+   */
+  struct Entry {
+    uint64_t hash;
+    Entry *next;
+    char data[0];
+
+    static uint32_t Size(uint32_t key_size, uint32_t value_size) {
+      return sizeof(Entry) + key_size + value_size;
+    }
+  };
+
+  /**
+   * An entry allocator
+   */
+  class EntryBuffer {
+   public:
+    /**
+     * Constructor for a buffer of entries.
+     *
+     * @param memory The memory pool to source memory for entries from
+     * @param entry_size The size of an entry
+     */
+    EntryBuffer(::peloton::type::AbstractPool &memory, uint32_t entry_size);
+
+    /**
+     * Destructor.
+     */
+    ~EntryBuffer();
+
+    /**
+     * Return a pointer to the next available Entry slot. If insufficient memory
+     * is available, memory is taken from the pool.
+     *
+     * @return A pointer to a free Entry slot.
+     */
+    Entry *NextFree();
+
+    /**
+     * Transfer all allocated memory blocks in this entry buffer into the target
+     * buffer.
+     *
+     * @param target Buffer where all blocks are transferred to
+     */
+    void TransferMemoryBlocks(EntryBuffer &target);
```

Comment:
About the memory blocks: when the entries are transfered from the thread local hash tables to the main hash table, only the pointer chain is moved, but the entries the still reside in the memory pool of the local threads - is that correct?

I think in this case all hash table instances use the same memory pool from the execution context [1], but maybe we should put an assertion here?

[1] https://github.com/tcm-marcel/peloton/blob/59d786081c80fca2f3b7d928330430f0a5745771/src/codegen/util/hash_table.cpp#L145

Reply:
Correct, they will use the same memory pool by design. But, I've added an assertion to be sure.